File: jci/pc.R
source('pc_wrapper.R', chdir=T)
library(methods)
library(pcalg)

set.seed(42)

# data: n x d matrix
dataset_raw <- read.csv(file='{FOLDER}{FILE}', sep=",", header=FALSE);
n <- dim(dataset_raw)[1]
d <- dim(dataset_raw)[2]

regimes <- read.csv(file='{FOLDER}{REGIMES}', sep=",", header=FALSE)

# number of context variables
r = max(regimes)
# p = |system var| + |context var|
p = d + r

context <- matrix(0, n, r)
for(i in 1:dim(regimes)[1]){
  regime <- regimes[i,1]
  if (regime != 0){
    context[i, regime] <- 1
  }
}

if({SKELETON}){
  fixedGaps <- read.csv(file='{FOLDER}{GAPS}', sep=",", header=FALSE)
  fixedGaps = (data.matrix(fixedGaps))
  rownames(fixedGaps) <- colnames(fixedGaps)
}else{
  fixedGaps = NULL
}

if({KNOWN}){
  targets <- read.csv(file='{FOLDER}{TARGETS}', sep=",", header=FALSE)
  targets <- (data.matrix(targets))
  show(targets)
}else{
  targets = NULL
}

dataset <- cbind(dataset_raw, context)
  
result <- pc_wrapper(data=dataset, systemVars=1:d, contextVars=(d+1):p, alpha='{ALPHA}', obsContext=matrix(0,1,r), test='{INDEP_TEST}', known='{KNOWN}', targets=targets)
show("pc.R")
show(result$cpdag)
dag <- as(pdag2dag(result$cpdag)[[1]], "matrix")
show(dag)

write.csv(dag, row.names=FALSE, file = '{FOLDER}{OUTPUT}');

File: jci/pc.py
"""
Modified from:
.. MIT License
..
.. Copyright (c) 2018 Diviyan Kalainathan
..
.. Permission is hereby granted, free of charge, to any person obtaining a copy
.. of this software and associated documentation files (the "Software"), to deal
.. in the Software without restriction, including without limitation the rights
.. to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.. copies of the Software, and to permit persons to whom the Software is
.. furnished to do so, subject to the following conditions:
..
.. The above copyright notice and this permission notice shall be included in all
.. copies or substantial portions of the Software.
..
.. THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.. IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.. FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.. AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.. LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.. OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
.. SOFTWARE.
"""
import os
import uuid
import warnings
import networkx as nx
from shutil import rmtree
from cdt.causality.graph.model import GraphModel
from pandas import DataFrame, read_csv
from cdt.utils.Settings import SETTINGS
from cdt.utils.R import RPackages, launch_R_script
import numpy as np
import torch


def message_warning(msg, *a, **kwargs):
    """Ignore everything except the message."""
    return str(msg) + '\n'

warnings.formatwarning = message_warning


class PC(GraphModel):
    """JCI-PC algorithm **[R model]**.

    Args:
        verbose (bool): Defaults to ``cdt.SETTINGS.verbose``.
    """

    def __init__(self, verbose=False):
        """Init the model and its available arguments."""
        if not RPackages.pcalg:
            raise ImportError("R Package pcalg is not available.")

        super().__init__()
        self.arguments = {'{FOLDER}': '/tmp/cdt_pc/',
                          '{FILE}': 'data.csv',
                          '{SKELETON}': 'FALSE',
                          '{GAPS}': 'fixedgaps.csv',
                          '{REGIMES}': 'regimes.csv',
                          '{TARGETS}': 'targets.csv',
                          '{VERBOSE}': 'FALSE',
                          '{ALPHA}': '1e-2',
                          '{OUTPUT}': 'result.csv'}
        self.verbose = SETTINGS.get_default(verbose=verbose)


    def _run_pc(self, data, fixedGaps=None, regimes=None, alpha=None,
                indep_test="gaussCItest", known=False, targets=None, verbose=True):
        """Setting up and running JCI-PC with all arguments."""
        id = str(uuid.uuid4())
        os.makedirs('/tmp/cdt_pc' + id + '/')
        self.arguments['{FOLDER}'] = '/tmp/cdt_pc' + id + '/'
        self.arguments['{ALPHA}'] = str(alpha)
        self.arguments['{INDEP_TEST}'] = indep_test
        if known:
            self.arguments['{KNOWN}'] = 'TRUE'
        else:
            self.arguments['{KNOWN}'] = 'FALSE'

        def retrieve_result():
            return read_csv('/tmp/cdt_pc' + id + '/result.csv', delimiter=',').values

        try:
            data.to_csv('/tmp/cdt_pc' + id + '/data.csv', header=False, index=False)
            if targets is not None:
                np.savetxt('/tmp/cdt_pc' + id + '/targets.csv', targets, delimiter=",")

            if fixedGaps is not None:
                fixedGaps.to_csv('/tmp/cdt_pc' + id + '/fixedgaps.csv', index=False, header=False)
                self.arguments['{SKELETON}'] = 'TRUE'
            else:
                self.arguments['{SKELETON}'] = 'FALSE'

            if regimes is not None:
                regimes.to_csv('/tmp/cdt_pc' + id + '/regimes.csv', index=False, header=False)
                self.arguments['{INTERVENTION}'] = 'TRUE'
            else:
                self.arguments['{INTERVENTION}'] = 'FALSE'

            pc_results = launch_R_script("{}/pc.R".format(os.path.dirname(os.path.realpath(__file__))), self.arguments, output_function=retrieve_result, verbose=verbose)
            print(pc_results)

        # Cleanup
        except Exception as e:
            rmtree('/tmp/cdt_pc' + id + '')
            raise e
        except KeyboardInterrupt:
            rmtree('/tmp/cdt_pc' + id + '/')
            raise KeyboardInterrupt
        rmtree('/tmp/cdt_pc' + id + '')
        return pc_results
